src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,315 @@
+            }
+
+            return true;
+        }
+
+        protected override string EscapeIdentifier(string identifier)
+            => identifier.EscapeIdentifier();
+
+        protected override async Task<SyntaxNode> ShouldPositionTriggerRefactoringAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var compilationUnit = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false) as CompilationUnitSyntax;  
+
+            // Here's conditions that trigger the refactoring (all have to be true in each scenario):
+            // 
+            // - There's only one namespace declaration in the document and all types are declared in it:
+            //    1. No nested namespace declaration (even it's empty).
+            //    2. The cursor is on the name of the namespace declaration.
+            //    3. The name of the namespace is valid (i.e. no errors).
+            //    4. No partial type declared in the namespace. Otherwise its multiple declaration will
+            //       end up in different namespace.
+            //
+            // - There's no namespace declaration and all types in the document are declared in global namespace:
+            //    1. The cursor is on the name of first declared type.
+            //    2. No partial type declared in the document. Otherwise its multiple declaration will
+            //       end up in different namespace.
+
+            var triggeringNode = GetTriggeringNode(compilationUnit, position);
+            if (triggeringNode != null)
+            {
+                var containsPartial = await ContainsPartialTypeWithMultipleDeclarationsAsync(document, triggeringNode, cancellationToken)
+                    .ConfigureAwait(false);
+
+                if (!containsPartial)
+                {
+                    return triggeringNode;
+                }
+            }
+
+            return default;
+
+
+            SyntaxNode GetTriggeringNode(CompilationUnitSyntax compUnit, int pos)
+            {
+                var namespaceDecls = compilationUnit.DescendantNodes().OfType<NamespaceDeclarationSyntax>().ToImmutableArray();
+
+                if (namespaceDecls.Length == 1 && compilationUnit.Members.Count == 1)
+                {
+                    var namespaceDeclaration = namespaceDecls.Single();
+                    Debug.Assert(namespaceDeclaration == compilationUnit.Members.Single());
+
+                    if (namespaceDeclaration.Name.Span.IntersectsWith(position)
+                        && namespaceDeclaration.Name.GetDiagnostics().All(diag => diag.DefaultSeverity != DiagnosticSeverity.Error))
+                    {
+                        return namespaceDeclaration;
+                    }
+                }
+                else if (namespaceDecls.Length == 0)
+                {
+                    var firstMemberDeclarationName = compilationUnit.Members.FirstOrDefault().GetNameToken();
```

Comment:
this looks like this would crash if there are no members in the file at all.  please add test for this.

Reply:
`GetNameToken` is an extension method that checks for null, but adding a test for this is a good idea.